cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tether/TetherServerHandler.java

```diff
@@ -0,0 +1,247 @@
+        lastMessageId = message.getId();
+      }
+    } catch (TopicNotFoundException e) {
+      LOG.warn("Received control connection from peer {} that's not tethered", peer);
+    }
+
+    if (lastMessageId != null) {
+      // Update the last message id for the topic if we read any messages
+      lastMessageIds.put(topic.getTopic(), lastMessageId);
+      String finalLastMessageId = lastMessageId;
+      TransactionRunners.run(transactionRunner, context -> {
+        AppMetadataStore.create(context).persistSubscriberState(topic.getTopic(), SUBSCRIBER, finalLastMessageId);
+      });
+    }
+
+    if (commands.isEmpty()) {
+      commands.add(new TetherControlMessage(TetherControlMessage.Type.KEEPALIVE, null));
+    }
+    Type type = new TypeToken<List<TetherControlMessage>>() { }.getType();
+    responder.sendJson(HttpResponseStatus.OK, GSON.toJson(commands, type));
+  }
+
+  /**
+   * Creates a tether with a client.
+   */
+  @POST
+  @Path("/tethering/connect")
+  public void createTether(FullHttpRequest request, HttpResponder responder)
+    throws NotImplementedException, IOException {
+    checkTetherServerEnabled();
+
+    String content = request.content().toString(StandardCharsets.UTF_8);
+    TetherConnectionRequest tetherRequest = GSON.fromJson(content, TetherConnectionRequest.class);
+    TopicId topicId = new TopicId(NamespaceId.SYSTEM.getNamespace(), "tethering_" + tetherRequest.getInstance());
+    try {
+      messagingService.createTopic(new TopicMetadata(topicId, Collections.emptyMap()));
+    } catch (TopicAlreadyExistsException e) {
+      LOG.warn("Topic {} already exists", topicId);
+    } catch (IOException e) {
+      LOG.error("Failed to create topic {}", topicId, e);
+      throw e;
+    }
+
+    // We don't need to keep track of the client metadata on the server side.
+    PeerMetadata peerMetadata = new PeerMetadata(tetherRequest.getNamespaces(), Collections.emptyMap());
+    // We don't store the peer endpoint on the server side because the connection is initiated by the client.
+    PeerInfo peerInfo = new PeerInfo(tetherRequest.getInstance(), null, TetherStatus.PENDING, peerMetadata);
+    store.addPeer(peerInfo);
+    responder.sendStatus(HttpResponseStatus.OK);
+  }
+
+  /**
+   * Accepts the tethering request.
+   */
+  @POST
+  @Path("/tethering/connections/{peer}/accept")
+  public void acceptTether(HttpRequest request, HttpResponder responder, @PathParam("peer") String peer)
+    throws NotImplementedException {
+    checkTetherServerEnabled();
+
+    PeerInfo peerInfo = store.getPeer(peer);
+    if (peerInfo.getTetherStatus() == TetherStatus.PENDING) {
+      store.updatePeer(peerInfo.getName(), TetherStatus.ACCEPTED);
+    } else {
+      LOG.info("Ignoring tethering accept as tethering state is {}", peerInfo.getTetherStatus());
+    }
+    responder.sendStatus(HttpResponseStatus.OK);
+  }
+
+  /**
+   * Rejects the tethering request.
+   */
+  @POST
+  @Path("/tethering/connections/{peer}/reject")
```

Comment:
Shall we make the states as query params?

Reply:
I tried making it a query param - /tethering/connections/{peer}/{action}
This doesn't work, fails with error:
```
java.util.regex.PatternSyntaxException: Illegal repetition near index 34
/v3/tethering/connections/([^/]+?)/{action})
                                  ^
	at java.util.regex.Pattern.error(Pattern.java:1957)
	at java.util.regex.Pattern.closure(Pattern.java:3159)
	at java.util.regex.Pattern.sequence(Pattern.java:2136)
	at java.util.regex.Pattern.expr(Pattern.java:1998)
	at java.util.regex.Pattern.compile(Pattern.java:1698)
	at java.util.regex.Pattern.<init>(Pattern.java:1351)
	at java.util.regex.Pattern.compile(Pattern.java:1028)
	at io.cdap.http.internal.PatternPathRouterWithGroups.add(PatternPathRouterWithGroups.java:95)cdapio
```

However, I did refactor common code into another method

Reply:
I meant `@QueryParam`.

Reply:
Ah, got it. fixed.